transformations/tag_subsequence_substitution/transformation.py

```diff
@@ -0,0 +1,159 @@
+import collections
+import itertools
+import random
+from typing import List, Tuple
+
+from interfaces.TaggingOperation import TaggingOperation
+from tasks.TaskTypes import TaskType
+from transformations.tag_subsequence_substitution.data import example_dataset
+
+"""
+A subsequence substitution--based augmentation system for tagging.
+Augmentation process:
+1. Draw a subsequence A from the input (tokens, tags) tuple.
+2. Draw a subsequence B within the whole dataset, with the same tag subsequence.
+3. Substitute A with B in the input example.
+
+Example:
+Input:
+    tokens: ['EU', 'rejects', 'German', 'call', 'to', 'boycott', 'British', 'lamb', '.']
+    (POS) tags: ['NNP', 'VBZ', 'JJ', 'NN', 'TO', 'VB', 'JJ', 'NN', '.']
+Drawn subsequence from input:
+    tokens: ['British', 'lamb']
+    (POS) tags: ['JJ', 'NN']
+Reference (randomly drawn from datset w.r.t. subsequence tags):
+    tokens: ['Germany', 'imported', '47,600', 'sheep', 'from', 'Britain', 'last', 'year', ',',
+             'nearly', 'half', 'of', 'total', 'imports', '.']
+    (POS) tags: ['NNP', 'VBD', 'CD', 'NN', 'IN', 'NNP', 'JJ', 'NN', ',', 'RB', 'NN', 'IN', 'JJ', 'NNS', '.']
+Substitution content:
+    tokens: ['last', 'year']
+    (POS) tags: ['JJ', 'NN']
+Output:
+    tokens: ['EU', 'rejects', 'German', 'call', 'to', 'boycott', 'last', 'year', '.']
+    (POS) tags: ['NNP', 'VBZ', 'JJ', 'NN', 'TO', 'VB', 'JJ', 'NN', '.']
+"""
+
+
+class TagSubsequenceSubstitution(TaggingOperation):
+    tasks = [TaskType.TEXT_TAGGING]
+    languages = "All"
+
+    def __init__(
+        self,
+        seed=0,
+        max_outputs=1,
+        base_dataset=example_dataset,
+        min_n_gram=1,
+        max_n_gram=4,
+    ):
+        super().__init__(seed, max_outputs=max_outputs)
+        assert (
+            base_dataset is not None
+        ), "base_dataset should be a list of (tokens, tags) tuples"
+        self.base_stats = self.extract_stats(
+            base_dataset, min_n_gram, max_n_gram
+        )
+        self.max_n_gram = max_n_gram
+        self.min_n_gram = min_n_gram
+
+    @staticmethod
+    def extract_stats(dataset, min_n_gram, max_n_gram):
+        stats = collections.defaultdict(collections.Counter)
```

Comment:
Do you think it makes sense to store these stats in a file here?

Reply:
Sure, I've added saving stats for potential reuse.